core/modules/date/date.tokens.inc

```diff
@@ -30,17 +32,108 @@ function date_token_info() {
   return $info;
 }
 
+/**
+ * Implements hook_token_info_alter().
+ */
+function date_token_info_alter(&$info) {
+  // Attach date field tokens to their respective entity tokens.
+  $field_info = field_info_fields();
+  $fields = _field_token_info();
+
+  foreach ($fields as $field_name => $field) {
+    if (in_array($field_info[$field_name]['type'], array('date', 'datetime', 'datestamp'))) {
+      foreach (array_keys($field['bundles']) as $token_type) {
+        // If a token already exists for this field, then don't add it.
+        if (isset($info['tokens'][$token_type][$field_name])) {
+          continue;
+        }
+  
+        // Ensure token type for this entity already exists. 
+        if (!isset($info['types'][$token_type]) || !isset($info['tokens'][$token_type])) {
+          continue;
+        }
+  
+        $info['tokens'][$token_type][$field_name] = array(
+          // Note that label and description have already been sanitized by _field_token_info().
+          'name' => $field['label'],
+          'description' => $field['description'],
+          'module' => 'date',
+          'type' => 'date-field-value',
+        );
+
+        if ($field_info[$field_name]['cardinality'] != 1) {
+          $info['tokens'][$token_type][$field_name]['dynamic'] = TRUE;
+          $info['tokens'][$token_type][$field_name]['description'] .= ' Replace the ? with the delta for multi-value fields.';
```

Comment:
Isn't that text supposed to be translatable (wrapped in t() )?

And also: if you need white space between translatable strings, it's better to not make it part of the translation. Too risky, that it's lost when translated.